### ISSUE TITLE
[filter-effects] Make grayscale() work without parameters

### ISSUE THREAD

**Issue by AmeliaBR** [w3c/fxtf-drafts#1]
As brought up by @svgeesus [on the mailing list](https://lists.w3.org/Archives/Public/public-fx/2016JanMar/0020.html):

The [ED of filters says that the lacuna value for the grayscale filter is zero](https://drafts.fxtf.org/filters/#FilterProperty).

However:
- the syntax does not seem to allow the argument to be omitted so how
  can the lacuna value be used
- if it did, then 0 is an odd value

0 means no change. Thus seems unintuitive. I would expect grayscale()
to be equivalent to grayscale(100%) which is likely what authors would
expect and would be convenient for the most common use case "make this
grayscale".

Suggested fix:
- change the grayscale filter syntax to make the argument optional
- change the lacuna value to 1

---

For what it's worth, this seems to be how -webkit-filter is implemented.
In other words, `-webkit-filter: grayscale()` applies 100% grayscale filter
(in Chrome anyway, haven't tested Safari).

For implementers of the standard property, neither Edge nor Firefox
currently support the function without a parameter, so changing the lacuna
value would not break anything that isn't already broken.

**Comment by dirkschulze** [comment]
@AmeliaBR Firefox does not support `grayscale()` and other filter functions without values. This probably should be brought up to the CSS WG but IIRC it was decided not to allow functions without values even though `-webkit-filter` does allow this. But I am not entirely sure. Maybe worth confirming.

**Comment by AmeliaBR** [comment]
@DirkSchulze If there's a good reason that default parameters for functions are not supported, then all the text about "lacuna values" should just be removed, as it's confusing.

Plus, as @tabatkins pointed out on the mailing list, [lacuna value is an undefined term from old SVG specs](https://lists.w3.org/Archives/Public/public-fx/2016JanMar/0077.html).

Since the phrase used by the spec is "lacuna value for interpolation", perhaps the main intent is to define the no-op version of each function in order to insert into a filter list.  Which makes perfect sense, and should be defined.  But it is not at all clear from the language used.

**Comment by dirkschulze** [comment]
@AmeliaBR If we require input values the default values are indeed superfluous. I am currently going through the minutes to get confirmation for a resolution on this topic.

**Comment by dirkschulze** [comment]
@dbaron @jwatt I am still looking through the minutes. Would Firefox be willing to support filter functions without function values? Note: This does not just affect `grayscale()` but all filter functions.

**Comment by mstange** [comment]
I don't see why not. Especially `grayscale()` makes a lot of intuitive sense without arguments.

**Comment by AmeliaBR** [comment]
I think a no-parameter version makes sense for

- `grayscale()` equivalent to `grayscale(100%)`
- `sepia()` equivalent to `sepia(100%)`
- `invert()` equivalent to `invert(100%)`

None of the others make sense without a parameter.  It would be confusing if these three were given maximum-effect parameter-less versions but the others were given no-op parameter-less versions.

The no-op versions for interpolation still need to be defined, but I recommend avoiding the term "lacuna value".  [CSS Transforms uses the phrase "identity transform function"](https://drafts.csswg.org/css-transforms/#interpolation-of-transforms), since these are functions that compute to the identity matrix; I'm not sure whether talking about an "identity filter function" helps anything.

**Comment by fsoder** [comment]
FTR, Blink has a use counter for the "no arguments" case:
https://www.chromestatus.com/metrics/feature/timeline/popularity/1403

Blink currently implements the "default value" for `grayscale()`, `sepia()` and `invert()` per @AmeliaBR's suggestion above, so there should be a decent chance for compat there. We could consider adding a counter for the other functions to see if that's a combination that authors use.

**Comment by longsonr** [comment]
Lacuna values are used when interpolating per https://www.w3.org/TR/filter-effects/#interpolation-of-filters. Changing the lacuna value would change how interpolation works in some cases. Firefox implements that specification text and uses lacuna values at that point. The default values are not superfluous even though in markup they are required.

**Comment by dirkschulze** [comment]
According to the Blink use counter quite a lot of ppl use "no argument" functions. As @longsonr mentions, the default values were introduced for interpolation.

Now I see the following problem:
`grayscale()` means `grayscale(1)` in WebKit/Blink
The lacuna value/initial value/default value (will open a different issue for the term) currently is `0` in the spec. For interpolation that makes a lot of sense and I'd be surprised if WebKit would not use `0` in this case CC @grorg @smfr . So we might end up defining a used value for interpolation and a different used value if no value was specified by the user.

**Comment by tabatkins** [comment]
Yes? That's completely fine. The "no-op" version of a function and the "default arguments" version of a function have no intrinsic connection to each other, and will naturally be different in cases where there is a "maximum" version of the effect.  (The "max" is probably what you want to default to when no arguments are specified.)

**Comment by AmeliaBR** [comment]
Yes, the default parameter should be completely different from the "value that is equivalent to `none`".  Otherwise, there's not much point in having a default parameter.

The no-effect values for each filter are only there for interpolation to `none`, and that needs to be clear.  The "lacuna value" term was confusing, since elsewhere that term was used for default values.  (Also, it's just a confusing, arbitrary jargon term in the first place.)

**Comment by dirkschulze** [comment]
@AmeliaBR The spec _does_ say initial value for interpolation.

I did. run some testing in Safari and Chrome. Both support value-less filter functions for all functions but `drop-shadow()`:
* `blur()`: `blur(0)` **no effect**
* `brightness()`: `brightness(0)`
* `contrast()`: `contrast(1)` **no effect**
* `drop-shadow()`: **not supported**
* `grayscale()`: `grayscale(1)`
* `hue-rotate()`: `hue-rotate(0)` **no effect**
* `invert()`: `invert(1)`
* `opacity()`: `opacity(1)` **no effect**
* `sepia()`: `sepia(1)`
* `saturate()`: `saturate(1)` **no effect**

Firefox follows the spec still.

`saturate()`, `contrast` don't seem to follow the general default values but IMO make sense. There is no clear "opposite".

IMO most values do make sense. I don't think that `brightness(0)` as default is useful since it simply turns elements to black. I rather would align them with `saturate()`, `contrast`.

@grorg @smfr Do you remember the thoughts that went into the default behavior?

**Comment by smfr** [comment]
I don't.

**Comment by dirkschulze** [comment]
Keeping "Needs WPT test" but close the issue.
